File: frontend/src/utils/safeRedirect.ts
type SafeRedirectOptions = {
  defaultPath?: string;
};

const DEFAULT_PATH = "/app";
const ALLOWED_HOSTS = ["kyradi.com"];

export function sanitizeRedirect(input: string | null, opts: SafeRedirectOptions = {}): string {
  const fallback = opts.defaultPath ?? DEFAULT_PATH;

  if (!input) {
    return fallback;
  }

  const trimmed = input.trim();
  if (!trimmed) {
    return fallback;
  }

  if (trimmed.startsWith("//")) {
    return fallback;
  }

  if (trimmed.startsWith("/")) {
    return trimmed;
  }

  try {
    const parsed = new URL(trimmed);
    const protocol = parsed.protocol.toLowerCase();
    if (protocol !== "http:" && protocol !== "https:") {
      return fallback;
    }

    const host = parsed.hostname.toLowerCase();
    const isAllowed =
      ALLOWED_HOSTS.includes(host) || ALLOWED_HOSTS.some((allowed) => host.endsWith(`.${allowed}`));

    return isAllowed ? trimmed : fallback;
  } catch {
    return fallback;
  }
}

File: backend/app/utils/safe_redirect.py
from urllib.parse import urlparse


def sanitize_redirect_url(value: str | None, default_path: str = "/app") -> str:
    if not value:
        return default_path

    trimmed = value.strip()
    if not trimmed:
        return default_path

    if trimmed.startswith("//"):
        return default_path

    if trimmed.startswith("/"):
        return trimmed

    parsed = urlparse(trimmed)
    if parsed.scheme not in {"http", "https"}:
        return default_path

    host = (parsed.hostname or "").lower()
    if host == "kyradi.com" or host.endswith(".kyradi.com"):
        return trimmed

    return default_path
